citations-tool/tool/src/java/org/sakaiproject/citation/tool/CitationHelperAction.java

```diff
@@ -1049,10 +1049,21 @@ protected void doPostJsonResponse(ParameterParser params,
 			String nestedCitations = params.getString("data");
 			if (nestedCitations!=null){
 
+				// Java has a bug where it throws a stackoverflow error when pattern matching very long strings
+				// http://bugs.java.com/view_bug.do?bug_id=5050507
+				// For very big lists, we just split the string
+				if (nestedCitations.length()>20000){
+					String[] parts = nestedCitations.split("\"sectiontype\"");
+					nestedCitations = "";
+					for (String part : parts) {
+						part = part.replaceAll("\\s+(?=([^\"]*\"[^\"]*\")*[^\"]*$)", "") + "\"sectiontype\"";  // replace whitespace (except between quotation marks) so can strip extra JSON arrays //  (it would be much better just to find a way of parsing the JSON without this string manipulation)
+						nestedCitations = nestedCitations + part;
+					}
+				}
+
 				// remove extra parentheses in json
 				// needed because of the extra ol's for accordion effect
 				nestedCitations = nestedCitations
-						.replaceAll("\\s+(?=([^\"]*\"[^\"]*\")*[^\"]*$)", "")    // replace whitespace
```

Comment:
I haven't escaped it for Java, all I did was make the inner group non-capturing, the original:
http://regexr.com/3bmvk

and the non-capturing:
http://regexr.com/3bmvn

the matching still looks the same. If it's ever the case that there is a stray quote though (eg at the end), then it does the opposite of what you are after:

http://regexr.com/3bmvq

Parsing with a regex is going to end in tears.